sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala

```diff
@@ -1259,57 +1259,33 @@ private[lf] object SBuiltinFun {
-                          coid,
-                          dstTplId,
-                          dstArg,
-                          allowCatchingContractInfoErrors = false,
-                        ) { dstContract =>
-                          // If the destination and src templates are the same, we skip the computation
-                          // of the destination template's view and the validation of the contract info.
-                          if (dstTplId == srcTplId)
-                            k(SAny(Ast.TTyCon(dstTplId), dstArg))
-                          else
-                            checkContractUpgradable(coid, srcContract, dstContract) { () =>
-                              executeExpression(machine, SEPreventCatch(dstView)) { dstViewValue =>
-                                if (srcViewValue != dstViewValue) {
-                                  Control.Error(
-                                    IE.Upgrade(
-                                      IE.Upgrade.ViewMismatch(
-                                        coid,
-                                        interfaceId,
-                                        srcTplId,
-                                        dstTplId,
-                                        srcView = srcViewValue.toUnnormalizedValue,
-                                        dstView = dstViewValue.toUnnormalizedValue,
-                                      )
-                                    )
-                                  )
-                                } else
-                                  k(SAny(Ast.TTyCon(dstTplId), dstArg))
-                              }
-                            }
-                        }
-                      }
+      resolvePackageName(machine, pkgName) { pkgId =>
+        val dstTplId = srcTplId.copy(packageId = pkgId)
+        machine.ensurePackageIsLoaded(
+          dstTplId.packageId,
+          language.Reference.Template(dstTplId),
+        ) { () =>
+          ensureTemplateImplementsInterface(machine, interfaceId, coid, dstTplId) {
+            fromInterface(machine, srcTplId, srcArg, dstTplId) {
+              case None =>
+                Control.Error(IE.WronglyTypedContract(coid, dstTplId, srcTplId))
+              case Some(dstArg) =>
+                getContractInfo(
+                  machine,
+                  coid,
+                  dstTplId,
+                  dstArg,
+                  allowCatchingContractInfoErrors = false,
+                ) { dstContract =>
+                  // If the destination and src templates are the same, we skip the computation
+                  // of the destination template's view and the validation of the contract info.
+                  if (dstTplId == srcTplId)
+                    k(SAny(Ast.TTyCon(dstTplId), dstArg))
+                  else
+                    checkContractUpgradable(coid, srcContract, dstContract) { () =>
+                      k(SAny(Ast.TTyCon(dstTplId), dstArg))
```

Comment:
If we compare this code to the one for `fetchTemplate`, then there are a bunch of non-obvious differences:
https://github.com/digital-asset/daml/blob/c6d90321727c2ec5cb700c73d6cc194f518b10d6/sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala#L2303-L2333

- Why don't we need the same validation vs. non-validation mode checks for `fetchInterface`?
- Why is there no activeness check for an interface? Could it be that an interface fetch of a contract archived in the same tx would wrongly succeed?

It seems these concerns could be alleviated by sharing at least the `importContract` code.

Reply:
We'll probably want a test for this as well - I'll coordinate with Paul on that, since it would fall under the test matrix.

Reply:
> We'll probably want a test for this as well - I'll coordinate with Paul on that, since it would fall under the test matrix.

"Probably?" ;-) 

Definitely! Ideally, the test would have been added as part of discovering the bug.

Reply:
Cool. That's great news then! Thanks for the thorough analysis.